EQL: HLRC documentation

astefan opened the pull request:
Adds HLRC documentation and fixes a bug in the HLRC client that prevented the response to return the `fields` section.

astefan committed:
HLRC documentation for EQL

astefan committed:
Formatting update

elasticmachine commented:
Pinging @elastic/es-docs (Team:Docs)

elasticmachine commented:
Pinging @elastic/es-ql (Team:QL)

astefan commented:
Link to the PR built docs: https://elasticsearch_80979.docs-preview.app.elstc.co/guide/en/elasticsearch/client/java-rest/7.16/java-rest-high-eql-eql-search.html

philkra commented:
@astefan could you, please add an example how the request is executed, for example how [Count API](https://elasticsearch_80979.docs-preview.app.elstc.co/guide/en/elasticsearch/client/java-rest/7.16/java-rest-high-count.html#java-rest-high-count-sync) does it?

bpintea commented on docs/java-rest/high-level/eql/search.asciidoc:
```diff
@@ -0,0 +1,71 @@
+<3> For basic queries, the maximum number of matching events to return.
+For sequence queries, the maximum number of matching sequences to return. Defaults to 10.
+<4> Field used to sort hits with the same timestamp in ascending order.
+<5> Field containing event timestamp. Defaults to `@timestamp`, as defined in the Elastic Common Schema (ECS).
```
nit, only if you apply any other changes: 
```suggestion
<5> Field containing the event timestamp. Defaults to `@timestamp`, as defined in the Elastic Common Schema (ECS).
```

bpintea reviewed (approved):
LGTM.
Wondering if linking the params to their [API equivalent](https://www.elastic.co/guide/en/elasticsearch/reference/master/eql-search-api.html) would make sense (not sure if that's a common practice for clients).

Luegg commented on docs/java-rest/high-level/eql/search.asciidoc:
```diff
@@ -0,0 +1,71 @@
+include-tagged::{doc-tests-file}[{api}-request]
+--------------------------------------------------
+<1> Comma-separated list of data streams, indices, or aliases used to limit the request.
+Supports wildcards (`*`). To search all data streams and indices, use `*` or `_all`.
```
maybe you want to mention the CCS capability as well?

Luegg reviewed (approved):
LGTM 👍 only few minor suggestions

costin reviewed (approved):
LGTM

costin commented on client/rest-high-level/src/main/java/org/elasticsearch/client/eql/EqlSearchResponse.java:
```diff
@@ -165,17 +170,27 @@ public int hashCode() {
+        private final Map<String, DocumentField> fetchFields;
 
-        public Event(String index, String id, BytesReference source) {
+        public Event(String index, String id, BytesReference source, Map<String, DocumentField> fetchFields) {
```
Since this is a public constructor, changing the signature breaks backwards compatibility for existing clients; the existing constructor needs to be kept in place while delegating to the new one.

astefan replied:
Agreed

costin commented on client/rest-high-level/src/main/java/org/elasticsearch/client/eql/EqlSearchResponse.java:
```diff
@@ -165,17 +170,27 @@ public int hashCode() {
                     return BytesReference.bytes(builder);
                 }
             }, SOURCE);
+            PARSER.declareObject(optionalConstructorArg(), (p, c) -> {
```
See my comment regarding the previous version which the new parser should still be able to understand.
As it stands right now, the new client breaks backwards compatibility. 
We expect only us to be able to create the Event objects however since the constructor is public (due to parsing) we need to be mindful of versioning issues.

astefan replied:
@costin the "previous" version of the server's EqlSearchResponse was sending the `fields` construct already, if it was requested by user. See https://github.com/elastic/elasticsearch/blob/7.16/x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/action/EqlSearchResponse.java#L287. But the old version of the HLRC eql search response parser [here](https://github.com/elastic/elasticsearch/blob/7.16/client/rest-high-level/src/main/java/org/elasticsearch/client/eql/EqlSearchResponse.java#L159-L168) apparently ignored the "unknown" construct for `fields` based on [this logic](https://github.com/elastic/elasticsearch/blob/7.16/libs/x-content/src/main/java/org/elasticsearch/xcontent/ObjectParser.java#L303-L304) and the parameter used to build the parser [here](https://github.com/elastic/elasticsearch/blob/7.16/client/rest-high-level/src/main/java/org/elasticsearch/client/eql/EqlSearchResponse.java#L155).

So, the new HLRC parser will understand the `fields` construct since now it is configured to parse the previously ignored `fields` construct sent by the server.

costin reviewed (changes requested):
The changes are not compatible with existing code; while Event class is normally created just by us, since the constructor is public other parties are able to access it.
Hence why we need to preserve compatibility or remote the public constructor all together.

astefan commented on client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/EqlDocumentationIT.java:
```diff
@@ -0,0 +1,122 @@
+        request.filter(QueryBuilders.matchAllQuery()); // <6>
+        request.resultPosition("head"); // <7>
+
+        List<FieldAndFormat> fields = new ArrayList<>();
+        fields.add(new FieldAndFormat("hostname", null));
+        request.fetchFields(fields); // <8>
```
I think we need to avoid the `.of` usage pattern for anything that's not 8.x, since all the previous ES versions should be compatible with 1.8 as well. And this PR targets 7.x versions only.

astefan committed:
Address reviews
Deprecate with the intention of limiting visibility of constructors and
methods that must not be public.

astefan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch '7.16' into eql_hlrc_documentation

costin reviewed (commented):
LGTM!
